src/CommandingOfficers/Cinder.java

```diff
@@ -192,76 +171,96 @@ protected void perform(MapMaster gameMap)
+    private WitchFireTracker tracker;
 
-    WitchFireAbility(Cinder commander)
+    WitchFireAbility(Cinder cinder)
     {
-      super(commander, WITCHFIRE_NAME, WITCHFIRE_COST);
-      coCast = commander;
+      super(cinder, WITCHFIRE_NAME, WITCHFIRE_COST);
     }
-
     @Override
-    protected void perform(MapMaster gameMap)
+    public void initForGame(GameInstance game)
     {
-      myCommander.addCOModifier(this);
+      tracker = WitchFireTracker.instance(game, WitchFireTracker.class);
     }
 
-    @Override // COModifier interface.
-    public void applyChanges(Commander commander)
+    @Override
+    protected void perform(MapMaster gameMap)
     {
-      coCast.witchFireListener.listen = true;
+      tracker.startTracking(myCommander);
     }
 
     @Override
-    public void revertChanges(Commander commander)
+    protected void revert(MapMaster gameMap)
     {
-      coCast.witchFireListener.listen = false;
+      tracker.stopTracking(myCommander);
     }
   }
 
-  private static class WitchFireListener implements GameEventListener
+  private static class WitchFireTracker extends StateTracker
   {
     private static final long serialVersionUID = 1L;
-    private Cinder myCommander = null;
-    private final int refreshCost;
-    public boolean listen = false;
 
-    public WitchFireListener(Cinder myCo, int HPCost)
+    private CountTracker<Commander, Unit> attackCounts = new CountTracker<>();
+
+    public void startTracking(Commander co)
+    {
+      attackCounts.getCountFor(co);
+    }
+    public void stopTracking(Commander co)
     {
-      myCommander = myCo;
-      refreshCost = HPCost;
+      attackCounts.resetCountFor(co);
     }
 
     @Override
     public GameEventQueue receiveBattleEvent(BattleSummary battleInfo)
     {
-      if( !listen )
+      Commander co = battleInfo.attacker.CO;
+      if( !attackCounts.hasCountFor(co) )
         return null;
+      // Since an active CO was part of the fight, reactivate the attacker at the cost of HP.
       GameEventQueue results = new GameEventQueue();
-      // Determine if we were part of this fight. If so, refresh at our own expense.
-      Unit minion = battleInfo.attacker;
-      if( minion.CO == myCommander )
+      Unit minion = battleInfo.attacker.unit;
+      // Cost starts at 1, then adds one for each subsequent attack
+      int refreshCost = 1+attackCounts.getCountFor(co, minion);
+      int hp = minion.getHP();
+      if( hp > refreshCost )
       {
-        int hp = minion.getHP();
-        if( hp > refreshCost )
-        {
-          minion.alterHP(-refreshCost);
-          minion.isTurnOver = false;
-        }
-        else
-        {
-          // Guess he's not gonna make it.
-          // TODO: Maybe add a debuff event/animation here as well.
-          Utils.enqueueDeathEvent(minion, results);
-        }
+        minion.alterHP(-refreshCost);
+        minion.isTurnOver = false;
+        attackCounts.incrementCount(co, minion);
```

Comment:
Isn't `attackCounts` a `StateTracker` that has been registered to receive events? It seems odd to be manually managing the counts from here.
Or alternatively: it seems weird to have a `StateTracker` to hold the counts.

Reply:
Actually, `CountTracker` doesn't even extend `StateTracker`, it's just in that package.

Reply:
Yes. We don't have a general `Util` namespace, so that's where I put it.